Update MCNP material reader to support default library identifier

mehmeturkmen opened the pull request:
This PR resolves the issue #1165. For now, the code requires a separate line or separate lines of default library naming. All the default library types were added, including nlib, hlib, elib and so on. 
Output: 
{1: pyne.material.Material({80160000: 0.118619, 922340000: 0.0006705, 922350000: 0.0846125, 922380000: 0.796098}, -1.0, 10.4, -1.0, {"mat_number":"1","table_ids":{"80160":{"hlib":"66h","nlib":"12c","plib":"12p"},"922340":{"hlib":"66h","nlib":"12c","plib":"12p"},"922350":{"hlib":"66h","nlib":"12c","plib":"12p"},"922380":{"hlib":"66h","nlib":"12c","plib":"12p"}}})}
 
"tests/mcnp_inp.tex" was modified but not "tests/test_mcnp.py" as some confusing lines. So it may need some changes for passing the test.

mehmeturkmen committed:
Update CHANGELOG.rst

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update mcnp_inp.txt

mehmeturkmen committed:
Update mcnp_inp.txt

gonuke commented:
Thanks @mehmeturkmen - this is a valuable addition, but I suggest adding a test for this.  You have updated the test input, but haven't added a test that tests for this new behavior, as far as I can tell.

(Note that the issue you reference is for an MCNP writer and not an MCNP reader, but this is useful nevertheless.)

gonuke commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,25 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+        # For deafult cross-section library. Requires a seperate line(s) of library indicator
+        # in material card 
+        lib_name = ['nlib', 'plib', 'hlib', 'pnlib', 'elib']
+        if line.split()[0][0:4] in lib_name:
```
I wonder if you could just loop through the entries in `line.split()` to see if any of them match?  This would allow the library indicator to be at any position, and also allow there to be multiple library indicators per line.

mehmeturkmen replied:
This modification works for any combination of multiple library indicators (including all in a single line) as long as any library indicator begins in a separate line or lines. I also tried to extend this modification through a mixture of library indicator and isotope indicator along with a fraction indicator but I did not achieve yet. As I understood from the function of `mat_from_inp_line`, searching for all possible cases needs a deep change in the considered function due to the way how it was written. But still there may be a short-cut. I will be working on it ! Besides the reader ability, I am going to add the writer capability for default library in this PR.

gonuke commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,25 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
         # make sure element/isotope is not commented out
-        if line.split()[0][0] != 'c' and line.split()[0][0] != 'C':
+        if line.split()[0][0] != 'c' and line.split()[0][0] != 'C' \
+                                     and (line.split()[0][0:4] not in lib_name):
```
If using my above plan, you'll need to set some flag rather than use this new test.

gonuke reviewed (changes requested):
I had a thought on allowing the library indicator to come at any point in the line.

gonuke commented:
I've updated the name of this PR so that we know what it is from the list. 

Also, best practice would be making a separate branch on your own fork for this rather than using the `0.7.0-rc` branch (even though it is on your own fork)

gonuke committed:
check for lib default anywhere on any line

mehmeturkmen committed:
Merge pull request #1 from gonuke/memo_mcnp

Suggested enhancement

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update test_mcnp.py

mehmeturkmen committed:
Update mcnp_inp.txt

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update test_mcnp.py

nsryan2 commented on CHANGELOG.rst:
```diff
@@ -28,6 +28,9 @@ v0.7.0 RC2
   * Functions to read OpenMC state point file to get meshtally data
 
+* MCNP Support
+  * Default library reading capability from material card, such as nlib, plib and etc.
```
```suggestion
  * Default library reading capability from material card, such as nlib, plib and etc. (#1269)
```

mehmeturkmen replied:
Done!

mehmeturkmen committed:
Merge branch '0.7.0-rc' into 0.7.0-rc

mehmeturkmen commented:
[x] Default library identifier reading capablity
[x] Test file was created for all possible combinations of library and atomic data
[x] Change log file conflict was resolved. 
[x] MCNP input file for testing was modified.

gonuke commented on CHANGELOG.rst:
```diff
@@ -27,7 +27,11 @@ v0.7.0 RC2
+* MCNP Support
+  * Default library reading capability from material card, such as nlib, plib and etc.
+
+* File scripts/activation_responses.py to get activation responses data for
+  ALARA output file for better visualization. Responses include:
```
```suggestion
* File scripts/activation_responses.py to get activation responses data for ALARA output file for better visualization. Responses include:
```
This needs to be one line to get the formatting right

gonuke commented on pyne/mcnp.py:
```diff
@@ -1425,9 +1441,12 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
             if len(data_string.split()[i].split('.')) > 1:
                 table_ids[str(zzzaaam)] = data_string.split()[i].split('.')[1]
+            else:
+                table_ids[str(zzzaaam)] = lib_id
```
These IDs are attributes of the whole material, not each nuclide.  For example, if someone manipulated the material to have addition nuclides (e.g. depletion/transmutation) we'd want the new nuclides to inherit this info.

They can be added as separate metadata, or as a stand alone entry in `table_ids`.  
E.g: `table_ids['default'] = lib_id` at some location that is not nuclide specfic, like when `table_ids` is first declared, above

Also, it should only be used if it's not empty.

mehmeturkmen replied:
I thought that to assign a default library identifier to each isotope would be more useful in the future pyNE releases. But, as you said, since all isotopes in a material card would use the default libraries, if exists, we can specify the lib as a separate metadata. So, your point definitely makes sense. In this case, I will try to add `lib_id` in `table_ids` with `default` key name in order to be more specific.

mehmeturkmen replied:
Solved!

gonuke commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,24 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+
         # make sure element/isotope is not commented out
         if line.split()[0][0] != 'c' and line.split()[0][0] != 'C':
+
```
Remove blank line (PEP8?)

gonuke commented on pyne/mcnp.py:
```diff
@@ -1417,6 +1428,11 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
     table_ids = {}
     for i in range(1, len(data_string.split())):
         if i & 1 == 1:
+
```
Remove blank line (PEP8?)

gonuke commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,24 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+    lib_name = ['nlib', 'plib', 'hlib', 'pnlib', 'elib']
     # people sometimes put comments in materials and then this loop breaks                                                                                       # so we need to keep reading if we encounter comments
     while len(line.split()) > 0 and (line[0:5] == '     ' or line[0].lower() == 'c'):
+        
```
Remove blank line (PEP8?)

gonuke commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,24 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
     while len(line.split()) > 0 and (line[0:5] == '     ' or line[0].lower() == 'c'):
+        
+        # This line reads deafult cross-section library from material card 
+        for token in line.split():
```
Why not do this at about 1443 where you are processing `data_string`?  This is a code maintainability question related to separation of concerns.  The overarching purpose of this loop is to filter out information that doesn't need to be processed.  The next loop will goes through this filtered string to process the necessary info.  It makes more sense to process the default lib info there.

mehmeturkmen replied:
Solved!

gonuke reviewed (changes requested):
Making good progress - I think this is all functional, but I have some style/maintainability and design requests.

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update CHANGELOG.rst

bam241 commented on pyne/mcnp.py:
```diff
@@ -1399,13 +1399,22 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+    lib_name = ['nlib', 'plib', 'hlib', 'pnlib', 'elib']
     # people sometimes put comments in materials and then this loop breaks                                                                                       # so we need to keep reading if we encounter comments
     while len(line.split()) > 0 and (line[0:5] == '     ' or line[0].lower() == 'c'):
+        # This line reads deafult cross-section library from material card 
```
```suggestion
        # This line reads default cross-section library from material card 
```

bam241 commented on pyne/mcnp.py:
```diff
@@ -1425,9 +1438,12 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
                 nucvec[zzzaaam] += float(data_string.split()[i+1])
             else:
                 nucvec[zzzaaam] = float(data_string.split()[i+1])
+            
```
```suggestion
```

bam241 commented on pyne/mcnp.py:
```diff
@@ -1417,6 +1426,10 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+            # Eliminates default library indicators
+            if data_string.split()[i].split('=')[0] in lib_name:
+                continue
+                
```
```suggestion

```

bam241 commented on pyne/mcnp.py:
```diff
@@ -1523,7 +1539,6 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
     return finished_mat
 
-
 class Wwinp(Mesh):
```
```suggestion

class Wwinp(Mesh):
```
I believe PEP8 requires 2 blank lines bewteen classes

bam241 reviewed (changes requested):
This looks good.

Some few minor formatting comments.

I'll let @gonuke merge that one as he has unresolved comments

mehmeturkmen commented:
@bam241 Do not merge yet. I am working on it. I am going to finalize it soon.

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update test_mcnp.py

mehmeturkmen committed:
Update CHANGELOG.rst

Co-authored-by: Nathan Ryan <55766947+nsryan2@users.noreply.github.com>

mehmeturkmen commented:
I think this PR now is done! Seems ready to merge.

mehmeturkmen committed:
Update pyne/mcnp.py

Co-authored-by: Baptiste Mouginot <15145274+bam241@users.noreply.github.com>

mehmeturkmen committed:
Update pyne/mcnp.py

Co-authored-by: Baptiste Mouginot <15145274+bam241@users.noreply.github.com>

mehmeturkmen committed:
Update pyne/mcnp.py

Co-authored-by: Baptiste Mouginot <15145274+bam241@users.noreply.github.com>

gonuke committed:
streamline finding of default libraries

mehmeturkmen committed:
Merge pull request #4 from gonuke/mcnp_nlib

streamline finding of default libraries

gonuke commented on pyne/mcnp.py:
```diff
@@ -1415,19 +1417,24 @@ def mat_from_inp_line(filename, mat_line, densities='None'):
+    table_ids['default'] = {}
     for i in range(1, len(data_string.split())):
         if i & 1 == 1:
+            # Eliminates default library indicators
```
I just realized that this won't work!!!  The logic of this `i & 1 == 1` threw me off

mehmeturkmen replied:
Still there may be a way to get over it! I am on it.

mehmeturkmen committed:
Update mcnp.py

I guess this is what was in your mind!

mehmeturkmen committed:
Update mcnp.py

mehmeturkmen committed:
Update test_mcnp.py

mehmeturkmen committed:
Merge pull request #5 from pyne/0.7.0-rc

0.7.0 rc

mehmeturkmen commented:
It seems we have a bunch of fails. I have rebased my branch so can we restart tests?

bam241 commented:
you need to rebase the branch so the test gets their latest version.

They are basically falling because they are out of date

gonuke committed:
Don't rely on odd index

This suggestion makes a number of changes.  In general, it is a little
more verbose with variables for long term clarity.  Specifically, it no
longer relies on the counting the entries in the token_list and acting
on the odd ones defined by (i & 1 == 1).  Instead, it pops tokens off
the list and acts on them according to their characteristics.  Some
passes through the while-loop only consume a single token from
the list while others consume 2 tokens from the list.

gonuke committed:
initialize defaul table_ids entry as dict

gonuke committed:
cleanup for PEP8

gonuke committed:
Use existing scheme

Apparently, there is already a scheme for storing and writing this
data from a Material object.

gonuke committed:
update test for prior scheme

gonuke committed:
get correct size for array of strings

gonuke committed:
don't get ahead of object creation

gonuke committed:
updated test to match old scheme

gonuke committed:
PEP8

gonuke committed:
manage case variations

gonuke committed:
tweak indenting (ignoring PEP8 for now)

gonuke committed:
PEP8 except indenting...

gonuke commented:
There is a new PR from me into your fork, @mehmeturkmen, that I think gets it all right.  It was a little more subtle than we expected.  I've also introduced some more variables to that it is more readable and maintainable in the long term.

I think you'll also need a rebase.

mehmeturkmen committed:
Merge pull request #7 from gonuke/mcnp_nlib

Don't rely on odd index

gonuke commented:
I think this needs a rebase still to update for new masses

gonuke commented:
This was absorbed in #1316 to finalize the expected results for the tests based on the updated mass data
